### DISCUSSION TITLE
Filtering basic auth credentials out of URI

### DISCUSSION THREAD

**Discussion by cilim** [vcr/vcr#887] [Ideas]
Hi everyone 👋 

I'm having trouble setting up VCR for APIs that require basic authentication credentials in the URI:
```
http://username:password@some-really-bad-api.com/
```

I don't think it's currently possible, but I might be wrong. These are my attempts:

```ruby
VCR.configure do |c|
  c.cassette_library_dir = "#{Rails.root}/spec/vcr_cassettes"
  c.hook_into :webmock

  c.filter_sensitive_data('username:password') do |interaction|
    next unless interaction.request.uri.include?('a-really-bad-api.com')
    
    interaction.request.uri.scan(/^https?:\/\/(.*)@a-really-bad-api-com/).first.first
  end
end
```

Doesn't work the 2nd time around when playing it back.

Then I tried this:
```ruby
  c.before_record do |interaction|
    next unless interaction.request.uri.include?('a-really-bad-api.com')
    
    safe_uri = interaction.request.uri.gsub(
      "#{Rails.application.secrets.bad_api_username}:#{Rails.application.secrets.bad_api_password}",
      "username:password"
    )
    
    interaction.request.uri = safe_uri
  end
  
  c.before_playback do |interaction|
    next unless interaction.request.uri.include?('a-really-bad-api.com')
    
    unsafe_uri = interaction.request.uri.gsub(
      /username:password/,
      "#{Rails.application.secrets.bad_api_username}:#{Rails.application.secrets.bad_api_password}"
    )
    
    interaction.request.uri = unsafe_uri
  end
```

Same thing...it wants to send the proper request, but the recorded cassette has `username:password` in it, so VCR raises the well-known error:
```
An HTTP request has been made that VCR does not know how to handle
```

Are there any plans for this feature? There was an issue a while back that was addressing this topic: https://github.com/vcr/vcr/issues/201

**Comment by emilebosch** [comment]
I think the the feature should be baked in to filter  _common_  (basic/bearer) credentials out by _default_ in the captured response. And we should give a clear message if people want to have them included.  For me that the credentials were included was a surprise to me.  

The reasoning is: _Its now too easy for a novice check in credentials  by accident into the repo. As a result need to scrub, rewrite the git repo, or in case of a breach get them exposed. I think most of the users will generate this data against a real API with real credentials._

The counter argument is that people should *know* about this, that recording is sensitive, which I agree with. But any breach prevented by having sane defaults saves people from silly accidents.

**Reply by krainboltgreene** [reply]
How does VCR know when you're recording sensitive data, in your mind? Bear in mind that your credentials may not be real, the recorded yaml can be entirely written by a human (and often is).

Further, even if we warn you, what is the path that you take when you decide not to record the credentails?

**Comment by krainboltgreene** [comment]
Lets talk about why this is difficult:

1. Secrets in a request can happen a ton of ways that are unpredictable, it's not just authN. The ceiling on this feature request is literally infinite.
2. Lets say you record a request with VCR that contains a secret. By some magic we know what piece of the request to "not record". Now you replay your recording, but you don't have that information anymore: It was never recorded. In your own example you encounter this: Part of VCR's identity detection (what request am I matching?) uses the URI which in your case contains the basic authentication. The thing you stored is different from the request generated.

The only way I see this being reasonably done is by encrypting recordings as needed like `ansible-vault`. This however considerably slows down any requests that do this. You can easily implement this yourself if needed.

**Reply by emilebosch** [reply]
Hi thanks for taking your time! I understand that you can't possibly cover all the use cases and that's okay, and that's not what I am not trying to advocate. However, one might probably lower the changes of accidental check ins (citation needed) and I think that in itself is valuable.

But maybe I am using VCR wrong? I use VCR to record API calls to external production systems, and replay them in my tests. In order not to pollute or stress these systems. These are either staging or production credentials, and our policy doesn't allow them to be checked in (which is good). 

And just like Rails scrubs certain fields from showing in the logs, I assumed VCR would do something similar for at least basic auth header and a bearer header when the yaml was produced for the things it could find.

Ofcourse for the use case where people write the YAML's by hand I'm not sure what would be the smart approach,  but we haven't used it like that (yet).

The first thing my and my other teams mates do after install the VCR gem is copying the sensitive data removal code. (see below)

I'm not sure if this is the way to go, but just thinking aloud:

- When serializing to yaml it checks for a set of auth headers `Authorization` etc. When it sees one of them in a block lists it doesn't serialise them. 
- If one doesn't want that, one can set a flag on a config to disable scrubbing or remove said header from a blocklist.
- A section on the readme

The code we're using is this:

```ruby
  VCR.configure do |c|
    c.cassette_library_dir = "test/fixtures/vcr"
    c.hook_into :webmock
    c.before_record do |i|
      i.response.headers.delete("Set-Cookie")
      i.request.headers.delete("Authorization")
      u = URI.parse(i.request.uri)
      i.request.uri.sub!(/:\/\/.*#{Regexp.escape(u.host)}/, "://#{u.host}")
    end
    c.register_request_matcher :anonymized_uri do |request_1, request_2|
      (URI(request_1.uri).port == URI(request_2.uri).port) &&
        URI(request_1.uri).path == URI(request_2.uri).path
    end
  end
  ```

**Comment by isaacbowen** [comment]
We're handling this by (a) filtering out Authorization header contents, and (b) using a custom header matcher that pays attention to everything _except_ "Authorization". Note that we're using `PLACEHOLDER_X` as our replacement pattern, not `<X>` or something, because VCR was trying to re-parse the substituted URI, and URI.parse doesn't like `https://user:<X>@host/path`.

[edit] Updating this sample to include the way we filter out env-configured API keys, because that feels useful too. :D

```ruby
VCR.configure do |config|
  # keep api keys out of recordings. we find names from what's in the environment,
  # but also (since CI probably doesn't have all (or any) of those configured) from
  # .env.default, which defines the things that a live environment is expected to
  # see, and which therefore is useful in predicting what our specs will be working with
  env_var_names_to_filter = ENV.keys.select { |key| key.ends_with?('_KEY') }
  env_var_names_to_filter += File.read(Rails.root.join('.env.default')).scan(/^\w+_KEY(?=)/m)

  env_var_names_to_filter.each do |key|
    # keeping this lookup dynamic to allow ENV to change across tests
    config.filter_sensitive_data("PLACEHOLDER_#{key}") { ENV[key] }
  end

  # assume authorization headers are sensitive
  config.filter_sensitive_data('PLACEHOLDER_AUTHORIZATION') do |interaction|
    next unless interaction.request.headers['Authorization']
    interaction.request.headers['Authorization'][0]
  end

  # a custom request matcher that looks at all headers *except* Authorization,
  # because we're filtering out the content of that header, which means
  # we can't use that header for matching
  headers_without_authorization = lambda do |request_1, request_2|
    request_1.headers.except('Authorization') == request_2.headers.except('Authorization')
  end

  config.default_cassette_options = {
    match_requests_on: [
      :method,
      :uri,
      :body,
      headers_without_authorization,
    ],
  }
end
```
